src/Diagnostics.DataProviders/DataProviders.cs

```diff
@@ -18,27 +20,85 @@ public class DataProviders
         public Func<MdmDataSource, IMdmDataProvider> Mdm;
         public Func<GenericMdmDataProviderConfiguration, IMdmDataProvider> MdmGeneric;
 
+        private readonly Dictionary<object, IMetadataProvider> _dataProvidersCache = new Dictionary<object, IMetadataProvider>();
+
         public DataProviders(DataProviderContext context)
         {
-            Kusto = new KustoLogDecorator(context, new KustoDataProvider(_cache, context.Configuration.KustoConfiguration, context.RequestId, context.KustoHeartBeatService));
-            Observer = new ObserverLogDecorator(context, SupportObserverDataProviderFactory.GetDataProvider(_cache, context.Configuration, context));
-            GeoMaster = new GeoMasterLogDecorator(context, new GeoMasterDataProvider(_cache, context));
-            AppInsights = new AppInsightsLogDecorator(context, new AppInsightsDataProvider(_cache, context.Configuration.AppInsightsConfiguration));
-            ChangeAnalysis = new ChangeAnalysisLogDecorator(context, new ChangeAnalysisDataProvider(_cache, context.Configuration.ChangeAnalysisDataProviderConfiguration, context.RequestId, context.clientObjectId, context.clientPrincipalName, Kusto, context.receivedHeaders));
-            Asc = new AscLogDecorator(context, new AscDataProvider(_cache, context.Configuration.AscDataProviderConfiguration, context.RequestId, context));
-            Mdm = (MdmDataSource ds) =>
+
+            Kusto = FromMemoryCache("Kusto", context) as IKustoDataProvider;
```

Comment:
Use this as an example if you decide to use the `GetOrAddDataProvider`
            Kusto = GetOrAddDataProvider(new KustoLogDecorator(context, new KustoDataProvider(
                _cache,
                context.Configuration.KustoConfiguration,
                context.RequestId,
                context.KustoHeartBeatService)));